File: Graphs/25/bellman.js
/**
 * @param {number[][]} times
 * @param {number} n
 * @param {number} k
 * @return {number}
 */
var networkDelayTime = function (times, n, k) {
  const distances = new Array(n).fill(Infinity);
  distances[k-1] = 0
  for (let i=0; i < n-1; i++) {
    let changed = false;
    for (const time of times) {
      const [u, v, w] = time;
      if (distances[u-1] + w < distances[v-1]) {
        distances[v - 1] = distances[u - 1] + w;
        changed = true;
      }
    }
    if (!changed) break;
  }
  const value = Math.max(...distances);
  return value === Infinity? -1 : value;
};

// Time Complexity: O(n*E)?

File: Graphs/25/bellman.py
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        INFINITY = float('inf')
        distances = [INFINITY] * n
        distances[k-1] = 0
        for i in range(n-1):
            changed = False
            for time in times:
                u, v, w = time
                if distances[u-1] + w < distances[v-1]:
                    distances[v-1] = distances[u-1] + w
                    changed = True
            if not changed:
                break
        value = max(distances)
        return -1 if value == INFINITY else value
